service/src/main/java/org/cbioportal/service/impl/CoExpressionServiceImpl.java

```diff
@@ -176,20 +176,19 @@
         // If the MolecularAlteration is for the query gene/geneset, skip it.  Otherwise,
         // filter out genetic_alteration values from genetic_alteration.VALUES
         // by considering oly the indices of the samples in the user query.
-        List<CoExpression> toReturn = new ArrayList<>();
+        List<CompletableFuture<CoExpression>> returnFutures = new ArrayList<>();
         for (MolecularAlteration ma : maItr) {
             String entityId = ma.getStableId();
             if (entityId.equals(queryGeneticEntityId)) {
                 continue;
             }
             List<String> internalValues = new ArrayList<>(Arrays.asList(ma.getSplitValues()));
             List<String> values = includedIndexes.stream().map(index -> internalValues.get(index)).collect(Collectors.toList());
-            CoExpression ce = computeCoExpressions(entityId, values, includedQueryValues, isMolecularProfileBOfGenesetType, threshold, molecularProfileId);
-            if (ce != null) {
-                toReturn.add(ce);
-            }
+            CompletableFuture<CoExpression> future = asyncMethods
+                .computeCoExpression(entityId, values, includedQueryValues, isMolecularProfileBOfGenesetType, threshold, molecularProfileId);
+            returnFutures.add(future);
         }
-        return toReturn;
+        return returnFutures.stream().filter(Objects::nonNull).map(CompletableFuture::join).collect(Collectors.toList());
```

Comment:
Couldn't you stream all of this, instead of starting off with a for loop? I think you have to remove the exception declarations from computeCoExpression, but it's not like we're handling those elegantly anyways.

```java
        return StreamSupport.stream(maItr.spliterator(), false)
            .filter(ma -> queryGeneticEntityId.equals(ma.getStableId()))
            .map(ma -> {
                String entityId = ma.getStableId();
                List<String> internalValues = new ArrayList<>(Arrays.asList(ma.getSplitValues()));
                List<String> values = includedIndexes.stream().map(index -> internalValues.get(index)).collect(Collectors.toList());
                return asyncMethods.computeCoExpression(entityId, values, includedQueryValues, isMolecularProfileBOfGenesetType, threshold, molecularProfileId);
            })
            .map(CompletableFuture::join)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());
```